services/s3/src/main/java/software/amazon/awssdk/services/s3/internal/crt/CrtContentLengthOnlyAsyncFileRequestBody.java

```diff
@@ -0,0 +1,55 @@
+/*
+ * Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License").
+ * You may not use this file except in compliance with the License.
+ * A copy of the License is located at
+ *
+ *  http://aws.amazon.com/apache2.0
+ *
+ * or in the "license" file accompanying this file. This file is distributed
+ * on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either
+ * express or implied. See the License for the specific language governing
+ * permissions and limitations under the License.
+ */
+
+package software.amazon.awssdk.services.s3.internal.crt;
+
+import java.nio.ByteBuffer;
+import java.nio.file.Path;
+import java.util.Optional;
+import org.reactivestreams.Subscriber;
+import org.reactivestreams.Subscription;
+import software.amazon.awssdk.annotations.SdkInternalApi;
+import software.amazon.awssdk.core.async.AsyncRequestBody;
+
+@SdkInternalApi
+public final class CrtContentLengthOnlyAsyncFileRequestBody implements AsyncRequestBody {
+    private final AsyncRequestBody asyncRequestBody;
+
+    public CrtContentLengthOnlyAsyncFileRequestBody(Path path) {
+        this.asyncRequestBody = AsyncRequestBody.fromFile(path);
+    }
+
+    @Override
+    public Optional<Long> contentLength() {
+        return asyncRequestBody.contentLength();
+    }
+
+    @Override
+    public void subscribe(Subscriber<? super ByteBuffer> subscriber) {
+        subscriber.onSubscribe(new Subscription() {
```

Comment:
It doesn't seem to comply with the spec
>Subscription.request and Subscription.cancel MUST only be called inside of its Subscriber context.

We can probably use [NoopSubscription](https://github.com/aws/aws-sdk-java-v2/blob/master/core/sdk-core/src/main/java/software/amazon/awssdk/core/internal/util/NoopSubscription.java) here. We need to move it outside of the internal package though since it's used cross module

```

            subscriber.onSubscribe(new NoopSubscription(s));
            subscriber.onError(new UnsupportedOperationException());
```

Alternatively, we can follow the same logic, i.e., calling onError in `request` and no-op in cancel.

Reply:
Will use `Alternatively` since its better to fail fast in such cases since its error if we try to subscribe CrtContentLengthOnlyAsyncFileRequestBody

Reply:
Oops, apologies, I mistyped, I meant we can follow the same logic in this class if we don't want to make NoopSubscription protected. We have to call onSubscribe first per reactive streams spec.